operator/src/main/java/oracle/kubernetes/operator/utils/ValidationUtils.java

```diff
@@ -0,0 +1,137 @@
+   *
+   * @param existingDomain domain that needs to be validated against.
+   * @param proposedDomain domain that needs to be validated.
+   * @return true if valid, otherwise false
+   */
+  public static boolean validateDomain(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    LOGGER.fine("Validating domain " + proposedDomain + " against " + existingDomain);
+    return unchanged(existingDomain, proposedDomain)
+        || validReplicas(proposedDomain)
+        || shouldIntrospect(existingDomain, proposedDomain);
+  }
+
+  private static boolean unchanged(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return existingDomain == proposedDomain || specEquals(existingDomain, proposedDomain);
+  }
+
+  private static boolean specEquals(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return Optional.of(existingDomain)
+        .map(Domain::getSpec)
+        .map(s -> isProposedSpecUnchanged(s, proposedDomain))
+        .orElse(false);
+  }
+
+  private static boolean isProposedSpecUnchanged(@NotNull DomainSpec existingSpec, @NotNull Domain proposedDomain) {
+    return existingSpec == proposedDomain.getSpec() || Objects.equals(existingSpec, proposedDomain.getSpec());
+  }
+
+  private static boolean shouldIntrospect(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return !Objects.equals(existingDomain.getIntrospectVersion(), proposedDomain.getIntrospectVersion())
+        || imagesChanged(existingDomain, proposedDomain);
+  }
+
+  private static boolean imagesChanged(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    if (!Objects.equals(existingDomain.getDomainHomeSourceType(), proposedDomain.getDomainHomeSourceType())) {
+      return true;
+    }
+    switch (proposedDomain.getDomainHomeSourceType()) {
+      case IMAGE:
+        return !Objects.equals(getImage(existingDomain), getImage(proposedDomain));
+      case FROM_MODEL:
+        if (noAuxiliaryImagesConfigured(existingDomain) && noAuxiliaryImagesConfigured(proposedDomain)) {
+          return !Objects.equals(getImage(existingDomain), getImage(proposedDomain));
+        } else {
+          return noAuxiliaryImagesConfigured(existingDomain)
+              || noAuxiliaryImagesConfigured(proposedDomain)
+              || !isEqualCollection(existingDomain.getAuxiliaryImages(), proposedDomain.getAuxiliaryImages());
+        }
+      default:
+        return false;
+    }
+  }
+
+  private static String getImage(@NotNull Domain existingDomain) {
+    return Optional.of(existingDomain).map(Domain::getSpec).map(DomainSpec::getImage).orElse(null);
+  }
+
+  private static boolean noAuxiliaryImagesConfigured(@NotNull Domain domain) {
+    return domain.getAuxiliaryImages() == null;
+  }
+
+  private static boolean validReplicas(@NotNull Domain domain) {
+    return Optional.of(domain)
+        .map(Domain::getStatus)
+        .map(DomainStatus::getClusters)
```

Comment:
I have found that it is easier to understand this sequence of steps when they are grouped on a single line, as they represent a single concept rather than multiple transformations:

```
.map(DomainStatus::getClusters).orElse(Collections.emptyList()).stream()
```

Alternately, you could extract the three steps into a single method named something like `getClustersAsStream`